crates/cubecl-opt/src/passes/array_copy_propagate.rs

```diff
@@ -0,0 +1,121 @@
+use std::collections::HashMap;
+
+use cubecl_core::ir::{Item, Operation, Operator, UnaryOperator, Variable};
+
+use crate::{AtomicCounter, Optimizer};
+
+use super::OptimizationPass;
+
+pub struct CopyPropagateArray;
+
+impl OptimizationPass for CopyPropagateArray {
+    fn apply_post_ssa(&mut self, opt: &mut Optimizer, changes: AtomicCounter) {
+        let arrays = find_const_arrays(opt);
+
+        for Array { id, length, item } in arrays {
+            let arr_id = id;
+            let vars = (0..length)
+                .map(|_| opt.root_scope.create_local_undeclared(item))
+                .collect::<Vec<_>>();
+            for var in &vars {
+                let local_id = opt.local_variable_id(var).unwrap();
+                opt.program.variables.insert(local_id, var.item());
+            }
+            replace_const_arrays(opt, arr_id, &vars);
+            changes.inc();
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+struct Array {
+    id: (u16, u8),
```

Comment:
Why two numbers for the id? Is it the variable id then the vectorization factor? Could it be 2 fields?

Reply:
It's the variable ID and depth. Since I'm working directly on the IR I need to mirror the ID scheme of the IR.